### PR TITLE
Add Apache Phoenix connector

### PR DESCRIPTION
**Pull request by vincentpoon**
This is a continuation of the work in #10536, which appears to have been abandoned .  The initial review comments there are addressed here.
In addition there's some code cleanup, and a couple new minor features: allow queries to hit secondary indexes, and support for CREATE/DROP schema.

A lot of the work was in getting the shaded phoenix-client right.  It's now under org.apache.phoenix and in an Apache repo here:
https://github.com/apache/phoenix-connectors/tree/4.x-HBase-1.4/presto-phoenix-shaded

### CODE REVIEW

**Comment by vincentpoon** [comment]
@kokosing Would be great if you could take a look, thanks!

@willshen

File: presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+                else if (typeAndValue.getType().equals(TimestampType.TIMESTAMP)) {
+                    statement.setTimestamp(i + 1, new Timestamp((long) typeAndValue.getValue()));
+                }
+                else if (typeAndValue.getType().equals(TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
+                    statement.setTimestamp(i + 1, new Timestamp(unpackMillisUtc((long) typeAndValue.getValue())));
+                }
+                else if (typeAndValue.getType() instanceof VarcharType) {
+                    statement.setString(i + 1, ((Slice) typeAndValue.getValue()).toStringUtf8());
+                }
+                else if (typeAndValue.getType() instanceof VarcharType) {
```
Comment (hustnn) [inline presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java]: Should be CharType here.

File: presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+                validType.equals(IntegerType.INTEGER) ||
+                validType.equals(DoubleType.DOUBLE) ||
+                validType.equals(RealType.REAL) ||
+                validType.equals(BooleanType.BOOLEAN) ||
+                validType.equals(DateType.DATE) ||
+                validType.equals(TimeType.TIME) ||
+                validType.equals(TimeWithTimeZoneType.TIME_WITH_TIME_ZONE) ||
+                validType.equals(TimestampType.TIMESTAMP) ||
+                validType.equals(TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE) ||
+                validType instanceof VarcharType ||
```
Comment (hustnn) [inline presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java]: adding validType instanceof CharType ||

File: presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java (hunk 1 above)
Comment (hustnn) [inline presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java]: duplicate VarcharType

File: presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java (hunk 2 above)
Comment (hustnn) [inline presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java]: Should support CharType

Comment (hustnn) [inline presto-phoenix/src/main/java/com/facebook/presto/plugin/phoenix/QueryBuilder.java]: Adding support for CharType
`validType instanceof CharType ||`

**Comment by vincentpoon** [comment]
Thanks @hustnn , made the changes you suggested

**Comment by hustnn** [comment]
I have one question. getQueryPlan is called twice. one to getSplits information and one to read the real data for each split. Is there any possibility that the scans for the two query plan are different?

**Comment by vincentpoon** [comment]
Yea @hustnn I think you're right, the data could change in between getting the splits information, and when executing the splits.  I think I can serialize the scans so the splits have the exact same scans from the query planner, which is the way the current Phoenix MapReduce framework works.  Thanks!

**Comment by hustnn** [comment]
@vincentpoon  I see. Cool. I will also try it from my side.

**Comment by vincentpoon** [comment]
@hustnn Pushed a commit that serializes the scans for each split.  It basically uses the Phoenix MR framework, where the scans are sent to each mapper.

**Comment by vincentpoon** [comment]
Pushed some fixes - split wasn't getting serialized properly, and there were some classloader issues.

**Comment by hustnn** [comment]
@vincentpoon 

I have the "Class org.apache.hadoop.net.StandardSocketFactory not found" during connection.

**Comment by vincentpoon** [comment]
@hustnn odd, did you try using my latest push?  The prior push had that error, but I fixed it in PhoenixConnectorFactory using ClassLoaderSafeConnectorSplitManager.

**Comment by JamesRTaylor** [comment]
Thanks for continuing the work, @vincentpoon. Would be much appreciated if we could get a review when you have some spare cycles, @kokosing.

**Comment by hustnn** [comment]
@vincentpoon .

When did you redeployed the presto-phoenix-shaded artifact?

Is the version 4.14.1-HBase-1.4 or 4.15.0-HBase-1.4-SNAPSHOT? It seems that the deployed artifact is still the old version.
https://mvnrepository.com/artifact/org.apache.phoenix/presto-phoenix-client-shaded/4.14.1-HBase-1.4

**Comment by hustnn** [comment]
@vincentpoon .

When I run the unit test, I still have the class not found error like this "java.lang.NoClassDefFoundError: org/apache/hadoop/hdfs/MiniDFSCluster".

I found this class is inside hadoop-hdfs-tests.jar, do you know how to include it? Do I need to add the dependency in presto-phoenix-shaded artifact?

**Comment by hustnn** [comment]
@vincentpoon A little strange, even build the shaded with your provided repo, still has the "java.lang.NoClassDefFoundError: org/apache/hadoop/hdfs/MiniDFSCluster" problem.

**Comment by hustnn** [comment]
@vincentpoon . As talked above, the RegionSizeCalculator.getRegionSize takes long time to calculate the region size. Now, I use the another way to only serialize the hbase scans, it avoid the long waiting time for region size calculation.

**Comment by vincentpoon** [comment]
@hustnn MiniDFSCluster actually wasn't in the jars, but it doesn't seem to be needed as I am able to run the tests, and the autobuild here ran it successfully also.  But anyways, I added to the phoenix-connectors repo I linked to above.  It's currently in the release pipeline, but you can install it yourself and update the pom here to "4.14.1-HBase-1.4-PPS-0.2".

Can you send me the diff of your commit?  I can add it here.  Thanks!

**Comment by hustnn** [comment]
@vincentpoon .
For the dependency problem, it is a little strange. I will try again.

Sure, I am cleaning the code. Will send the diff to you next week.

**Comment by vincentpoon** [comment]
@hustnn I updated the pom with the latest shaded jar version.  See if it works for you now

**Comment by vincentpoon** [comment]
@hustnn I can just take the code from PhoenixInputFormat for now without the region sizing, we can make an upstream change later in Phoenix to avoid the code duplication.  WDYT?

**Comment by vincentpoon** [comment]
@hustnn I removed the region size calculation.  We were also finding the region locations twice, i fixed that also.
FYI @lhofhansl

**Comment by lhofhansl** [comment]
Nice. @vincentpoon .

I had a question too... Is there a way to pass through query hints to Phoenix, such a NO_CACHE, or NO_INDEX?

For analytical queries the HBase blockcache is not always helpful (or at least you might not want to pollute that cache with full scans - and rather keep the cache for OLTP-type queries).

**Comment by lhofhansl** [comment]
It's worth doing, I think.
Perhaps later with separate PR.
I played with this PR. And it works pretty well.

**Comment by lhofhansl** [comment]
I was wondering about what would be the best way to proceed on this? This could and should be the connector for HBase (via Phoenix).

Should we maintain this as part of presto like many of the other connectors, or is this something that the should be maintained as part of Phoenix?

**Comment by aertoria** [comment]
@martint

**Comment by aertoria** [comment]
Talked with @martint this morning,  please carry on the p.r. to the new master https://github.com/prestosql/presto
@vincentpoon  @JamesRTaylor

**Pull request closed without merging**